Remove (deprecated) BTE Framework from 7.x codebase

abollini opened the pull request:
## References
* Fixes #2817
* Fixes #3012

## Description
All the references to the BTE framework has been removed from the code base. The extract metadata step has been modified to use the Live Import Framework and all the code related to enhancement of current workspace item both via identifiers than via file has been centralized in such step. This step is NOT part of the default configuration, users need to enable it explicitly.

Finally some cleanup to the submission steps hierarchy have been performed and the lifecycle of the patch/upload processing for inprogress items have been fixed so that all the steps participate in the preProcessing phase and, after than the target step has processed the request, they are involved again in the postProcessing.

## Instructions for Reviewers
Review the changes in the WorkspaceItemRestIT the test added cover the extract metadata step that previously was not tested.

To test manually the metadata-extract step you need to uncomment the step in the submission-forms.xml

adding a pmid to the dc.identifier.other should lead to the automatic enrichment of the item with information retrieved from PubMed.

Test the submission features at large to verify that the BTE removal has had no side effect (but the large number of ITs should make us safe)

## Checklist

- [X] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [X] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [X] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [X] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [N/A] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [N/A] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

abollini committed:
Remove BTE and implement the metadata extraction step using the Live Import

abollini committed:
Fix dependencies for dspace-rest (legacy)

abollini committed:
Align test to the configuration

tdonohue reviewed (changes requested):
@abollini : Gave this a review today... Overall, it looks good, and I'm glad to see tests are provided for the new code (the new, disabled-by-default, extraction step).  I found a few very minor issues that need cleanup.

I also tried manually testing the metadata extraction step....while the tests show that it *should* work, it doesn't appear to work from the UI.  Also, enabling it blocks the ability to complete a submission, as the UI shows this step as being required (and empty):
![extract-step](https://user-images.githubusercontent.com/483997/112036331-83161e80-8b0e-11eb-8eb7-76711e9e5981.png)

I also cannot seem to trigger the step to take action (I tried uploading a BibTex file, and also adding a PubMed ID to dc.identifier.other and neither worked)

I think we need to do one of the following with this PR:
1. Either we remove the Extraction Step code/configurations from this PR & create a follow-up PR with that work  (This allows us to move the BTE Framework removal forward and finish up the Extraction Step later)
OR
2. Merge it as-is, but add a warning that this Extraction Step does NOT currently work from the UI.  Then, we'd have to follow up later with UI fixes to allow the Extraction Step to work.

I lean slightly towards (1), as the Extraction Step is not required at this time (it's not in the 7.0 scope, instead it's related to https://github.com/DSpace/dspace-angular/issues/896 which is scheduled for 7.1), but the BTE removal *is required*.

mwoodiupui commented on dspace-api/pom.xml:
```diff
@@ -537,17 +537,17 @@
+		<dependency>
+			<groupId>net.sf.opencsv</groupId>
+			<artifactId>opencsv</artifactId>
+			<version>2.3</version>
```
See https://mvnrepository.com/artifact/net.sf.opencsv/opencsv where it says "Note: This artifact was moved to:  com.opencsv » opencsv"  The two have different `groupId` but are the same project, ten years (and two repositories) apart.  We should use at least 5.2.

abollini replied:
done

abollini committed:
Switch to opencvs in the com.opencsv groupId

abollini committed:
Fix type of the extract panel in the configuration

KevinVdV reviewed (changes requested):
I made an initial review of the code see comment above, aside from that comment I have the following issues/questions:

* The tests are currently making calls to the pubmed service itself, that should be mocked
* I noticed that you removed a couple of classes in the "org.dspace.submit.step" package in the dspace-api, is there a reason why the other ones aren't removed ? They don't appear to contain any code is there a reason to not remove that entire package ?
* I also couldn't get the extraction of metadata to work from the UI

abollini committed:
Cleanup submission code

Moved shared code among workspace and workflow repository to the submission service
introduced a base RestProcessingStep interface to be specialized among ListenerStep and AbstractRestProcessingStep (data processing)
removed retrieval of the Request from the service passing directly the http request that was available
added test to check the special case of steps that should be not exposed as panel (extract step)

abollini committed:
Make a common ancestor for the Listener and processing steps

abollini committed:
Fix misleading properties names for the CharacterSeparated provider

abollini committed:
Mock the publication import service for testing

abollini committed:
Code cleanup: remove unused submit code from dspace-api

abollini commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/ExtractMetadataStep.java:
```diff
@@ -9,105 +9,207 @@
         throws IOException {
 
         Item item = wsi.getItem();
+        File file = Utils.getFile(multipartFile, "extract-metadata-step", stepConfig.getId());
```
all the uploadable steps are invoked when a file is uploaded, the "extract-metadata-step" prefix is only used for the temporary files created for the step processing

abollini commented:
@tdonohue @KevinVdV all your feedback should have been addressed. I haven't tested yet with a running Angular UI we can do that tomorrow but I would appreciate if you are able to quickly retest and let me know. As the extract step is expected to be invisible to the UI (and now I have added a test to verify - and fix the previous mistake) everything should be ok. The PR is beome a bit larger as I have add to add the pubmed import service mock, extra tests and refactor other part of the submission code removing unused class and centralizing some duplicate code among WorkspaceItemRestRepository and WorkflowItemRestRepository

abollini commented on dspace-api/pom.xml:
```diff
@@ -537,17 +537,17 @@
+			<groupId>org.jbibtex</groupId>
+			<artifactId>jbibtex</artifactId>
+			<version>1.0.10</version>
+		</dependency>
```
done

abollini commented on dspace-api/src/test/data/dspaceFolder/config/item-submission.xml:
```diff
@@ -81,20 +82,16 @@
+        <step-definition id="extractionstep"> 
+            <heading>submit.progressbar.ExtractMetadataStep</heading>
+            <processing-class>org.dspace.app.rest.submit.step.ExtractMetadataStep</processing-class>
+            <type>cclicense</type> 
```
done

tdonohue commented on dspace/config/item-submission.xml:
```diff
@@ -119,14 +119,15 @@
+        <!--Step will be to enrich the current submission querying external providers or processing the uploaded file -->
+        <!-- Uncomment this step to enrich the current submission using information extracted 
+            from uploaded files or metadata. Please check also config/spring/api/step-processing-listener.xml
+            for further configuration -->
```
Just a note which we may want to add to this configuration.  It looks like, currently, this step is only triggered during either:
* An **upload** (e.g. extract metadata from a BibTex file)
* A **save** of the form. (e.g. entering a PubMed ID only triggers a lookup if you click "Save" or "Save for later")

abollini replied:
done. I have also mentioned the autosave feature available on the angular side that can be used with that to make the process more user friendly

tdonohue reviewed (approved):
👍 @abollini : This looks good to me now.  

* The BTE functionality is removed (and I'm not noticing any obvious side-effects)
* The new automated metadata extraction step is now invisible (not shown in the Submission form) and it's _disabled by default_ as [we recommended in ticket 3012](https://github.com/DSpace/DSpace/issues/3012#issuecomment-714730029)

I've also successfully tested the metadata extraction step from the Submission UI.  However, it's worth noting that it's only triggered during a file upload or saving of the form.  So, we should document that (see my inline note below), especially since the form no longer auto-saves by default.

In any case, this looks good to me.  Thanks!

abollini committed:
Add more information about the extraction step

tdonohue commented:
@KevinVdV : Could you give this another quick review?  It looks (to me) like your prior feedback has all been addressed.  But, I'd prefer if you took a second look before merger. Thanks!

KevinVdV reviewed (commented):
My issues with the code have been fixed, so on a code basis I can approve the PR. However when testing it I ran into some issues.

I enabled the "extractionstep" in the "item-submission.xml" by:
* Commenting out the "step-definition" section
* Adding the step in the "submission-process" section

I then started a new submission & added the "32373993" value to the dc.identifier.other field & pressed the "save" button. I then got a 500 error response code from the PATCH request.

I found the following exception in the dspace logs:

`
	at org.dspace.importer.external.service.components.AbstractRemoteMetadataSource.retry(AbstractRemoteMetadataSource.java:179) ~[dspace-api-7.0-beta5-SNAPSHOT.jar:7.0-beta5-SNAPSHOT]
	at org.dspace.importer.external.pubmed.service.PubmedImportMetadataSourceServiceImpl.getRecord(PubmedImportMetadataSourceServiceImpl.java:132) ~[dspace-api-7.0-beta5-SNAPSHOT.jar:7.0-beta5-SNAPSHOT]
	at org.dspace.external.provider.impl.LiveImportDataProvider.getExternalDataObject(LiveImportDataProvider.java:85) ~[dspace-api-7.0-beta5-SNAPSHOT.jar:7.0-beta5-SNAPSHOT]
	at org.dspace.submit.listener.SimpleMetadataListener.getExternalDataObject(SimpleMetadataListener.java:62) ~[dspace-api-7.0-beta5-SNAPSHOT.jar:7.0-beta5-SNAPSHOT]
	at org.dspace.app.rest.submit.step.ExtractMetadataStep.doPostProcessing(ExtractMetadataStep.java:97) ~[classes/:?]
	... 131 more
Caused by: org.dspace.importer.external.exception.MetadataSourceException: At retry of operation 47270d87-e605-49db-8f92-6d706224eef3 0
	at org.dspace.importer.external.service.components.AbstractRemoteMetadataSource.throwSourceException(AbstractRemoteMetadataSource.java:248) ~[dspace-api-7.0-beta5-SNAPSHOT.jar:7.0-beta5-SNAPSHOT]
	at org.dspace.importer.external.service.components.AbstractRemoteMetadataSource.retry(AbstractRemoteMetadataSource.java:168) ~[dspace-api-7.0-beta5-SNAPSHOT.jar:7.0-beta5-SNAPSHOT]
	at org.dspace.importer.external.pubmed.service.PubmedImportMetadataSourceServiceImpl.getRecord(PubmedImportMetadataSourceServiceImpl.java:132) ~[dspace-api-7.0-beta5-SNAPSHOT.jar:7.0-beta5-SNAPSHOT]
	at org.dspace.external.provider.impl.LiveImportDataProvider.getExternalDataObject(LiveImportDataProvider.java:85) ~[dspace-api-7.0-beta5-SNAPSHOT.jar:7.0-beta5-SNAPSHOT]
	at org.dspace.submit.listener.SimpleMetadataListener.getExternalDataObject(SimpleMetadataListener.java:62) ~[dspace-api-7.0-beta5-SNAPSHOT.jar:7.0-beta5-SNAPSHOT]
	at org.dspace.app.rest.submit.step.ExtractMetadataStep.doPostProcessing(ExtractMetadataStep.java:97) ~[classes/:?]
	... 131 more
Caused by: java.lang.RuntimeException: java.lang.ClassNotFoundException: com.sun.ws.rs.ext.RuntimeDelegateImpl
`

It appears to be an issue with a library somewhere ?

Tomcat I used: 9.0.29
Java version: 11.0.2

mwoodiupui commented:
That class should have been found in jersey-common.

tdonohue commented:
@abollini : There's a new small merge conflict here...I think it was caused by a minor change to the dspace-api/pom.xml in #3171 (which was just merged).

In the meantime, I'll see if I can reproduce the dependency issue that @KevinVdV [noted above](https://github.com/DSpace/DSpace/pull/3189#pullrequestreview-628720183).

mwoodiupui commented:
Since it appears to be trying date formats one after another until one succeeds, perhaps those INFOs should be DEBUGs.

abollini committed:
Merge remote-tracking branch 'origin/main' into removeBTE

abollini committed:
Improve logging of date parsing issues

abollini commented:
I have resolved the merge conflict and take the chance to improve the logging as by @mwoodiupui suggestion in this commit https://github.com/DSpace/DSpace/pull/3189/commits/683cba643efa2acca27569bb988878fb72b66817
(if all the parser fail a single log info is now reported)

abollini commented:
I have also performed a new manual test, similar to what @tdonohue performed and I confirm that everything works as expected. I'm not able to reproduce the issue reported by @KevinVdV

KevinVdV reviewed (approved):
I tried to test the extraction step on another machine & it worked there. Most be an issue with my local setup. Not going to hold this PR on that. Approving.

tdonohue commented:
Thanks @abollini !  Merging this as it's now at +2.
